### ISSUE TITLE
Wrath Classic: Edge cases

### ISSUE THREAD

**Issue by Road-block** [Oppzippy/TankMD#9]
Very nice addon, there are however a couple issues.

Not bugs that cause Lua errors but logic bugs that cause unexpected behavior.

For one `GetSortedGroupMembers` in TargetMatcher is unreliable in that it would produce a table that's incomplete in raids with empty raid ids.
Furthermore iterating with ipairs would stop at the first non-consecutive integer key.

A better/safer implementation would be something like this
```lua
function TargetMatcherPrototype:GetSortedGroupMembers()
	local groupMembers = {}
	local units = IsInRaid() and self.raidUnits or self.partyUnits
	local grouplimit = IsInRaid() and MAX_RAID_MEMBERS or MAX_PARTY_MEMBERS+1
	for i = 1, grouplimit do
		local unit = units[i]
		local name = UnitName(unit)
		if name and name ~= UNNOWNOBJECT then
			groupMembers[#groupMembers+1] = name
		end
	end

	table.sort(groupMembers)

	return groupMembers
end
```
I don't understand why groupmembers are sorted alphabetically but not important.

A second problem is that while this probably works ok for machine formed 5mans (like from RDF) where you can only have 1 tank and 1 healer it is unreliable in raids and even manually formed 5mans.
In raids you often have members with `UnitGroupRolesAssigned` "TANK" or "HEALER" that do not actually fulfill that role.
In PuGs it's even unrealistic to ask everyone to set their role properly or the RL to do it for them.

In combination with the targets being returned in alphabetical order I have often had the addon keep misdirecting to the first name with a role of "TANK" regardless if they were set as a "MAINTANK" or not.

There is no perfect solution to this but I would prefer if the addon had a stricter match to raid role in raids.
For example something like this in RoleTargetMatcher.
```lua
function RoleTargetMatcherPrototype:Matches(unit)
	local role = UnitGroupRolesAssigned(unit)
	local raidTank = GetPartyAssignment("MAINTANK",unit)
	if role ~= "NONE" then
		if IsInRaid() then
			return (self.role == "TANK") and raidTank or (self.role == role)
		else
			return self.role == role
		end
	elseif LGIST then
		local guid = UnitGUID(unit)
		local inspectInfo = LGIST:GetCachedInfo(guid)
		if inspectInfo then
			return self.role == inspectInfo.spec_role
		end
	end
end
```
As a hunter I'd prefer if I mistakenly misdirect to my pet, rather than to a dps player that has not set their roles correctly, essentially griefing them.

A more nuanced solution would be to have the addon _fallback_ to grouprole when no matching raidrole is found, but that would require a more extensive refactoring.

Thanks for your time :)

**Comment by Oppzippy** [comment]
> For one GetSortedGroupMembers in TargetMatcher is unreliable in that it would produce a table that's incomplete in raids with empty raid ids.

I didn't think this was possible. You're suggesting it's possible for raid ids to be skipped (ex: raid1 present, raid2 absent, raid3 present)? These numbers do not correspond to raid group and position in the raid group, if that is the cause of the confusion.

>  Furthermore iterating with ipairs would stop at the first non-consecutive integer key.

Shouldn't be a problem since keys are always consecutive, unless maybe UnitName can return nil for a valid unit in some situation?

> ```lua
>     local name = UnitName(unit)
>     if name and name ~= UNNOWNOBJECT then
> ```

I remember many years ago seeing that bug where people have the name Unknown, but I thought that was long gone. I haven't played WoW in over two years though, so I could be misremembering. Does this still occur, and if so, is it common enough to matter?

> I don't understand why groupmembers are sorted alphabetically

This is so that ordering of tanks is predictable rather than based on whatever id between raid1 and raid40 happens to get assigned to them.

> A second problem is that while this probably works ok for machine formed 5mans (like from RDF) where you can only have 1 tank and 1 healer it is unreliable in raids and even manually formed 5mans.

Yeah, I originally made this addon for mythic+, so I did the bare minimum for raids. Having some way of displaying to the user what tank number corresponds to which player is an example of a missing raid feature that I just never bothered with.

> In combination with the targets being returned in alphabetical order I have often had the addon keep misdirecting to the first name with a role of "TANK" regardless if they were set as a "MAINTANK" or not.
> A more nuanced solution would be to have the addon fallback to grouprole when no matching raidrole is found, but that would require a more extensive refactoring.

I think this should probably be a setting, with the options: "Ignore Main Tank Assignment", "Include Main Tanks", "Prioritize Main Tanks", and "Main Tanks Only". This gives as many options as possible to work with weird setups. I'll implement this when I get a chance.

**Comment by Road-block** [comment]
Unit methods returning "empty" information is still possible around load times.

Holes in raid rosters are - also not common but - possible.
That's the source of notes like on <https://warcraft.wiki.gg/wiki/API_GetRaidRosterInfo> 

Options for favoring raid assignments vs role etc sound good :)

The hardcoded prio to raid tanks works for me, so no rush :)

**Comment by Oppzippy** [comment]
`GetSortedGroupMembers` should be fixed in the latest alpha.

**Comment by Oppzippy** [comment]
This was fixed in v3.0.0 and I forgot to close the issue.
